File: javascript/ocr-numbers/ocr-numbers.js
const DIGITS = [
    [' _ ',
     '| |',
     '|_|'],
    ['   ',
     '  |',
     '  |'],
    [' _ ',
     ' _|',
     '|_ '],
    [' _ ',
     ' _|',
     ' _|'],
    ['   ',
     '|_|',
     '  |'],
    [' _ ',
     '|_ ',
     ' _|'],
    [' _ ',
     '|_ ',
     '|_|'],
    [' _ ',
     '  |',
     '  |'],
    [' _ ',
     '|_|',
     '|_|'],
    [' _ ',
     '|_|',
     ' _|']
].map((a) => a.join(''));

function convertDigit(digit) {
    const digitIdx = DIGITS.indexOf(digit.join(''));
    return digitIdx != -1 ? digitIdx.toString() : '?';
}

export function convert(input) {
    const output = [];
    const lines = input.split('\n');
    let topLine = 0;
    while (lines.length > topLine) {
        if (topLine != 0) {
            output.push(',');
        }
        let leftColumn = 0;
        while (lines[topLine].length > leftColumn) {
            output.push(convertDigit([
                lines[topLine].slice(leftColumn, leftColumn + 3),
                lines[topLine + 1].slice(leftColumn, leftColumn + 3),
                lines[topLine + 2].slice(leftColumn, leftColumn + 3)
            ]));
            leftColumn += 3;
        }
        topLine += 4;
    }
    return output.join('');
}

File: ruby/ocr-numbers/_iterations/3/ocr_numbers.rb
module OcrNumbers
  def self.convert(input)
    input.lines.map(&:chomp).each_slice(LINES_PER_GROUP).map do |lines|
      convert_group(lines)
    end.join(",")
  end

  private

  LINES_PER_GROUP = 4
  COLUMNS_PER_DIGIT = 3
  NUMBERS = [
    [" _ ",
     "| |",
     "|_|",
     "   "],
    ["   ",
     "  |",
     "  |",
     "   "],
    [" _ ",
     " _|",
     "|_ ",
     "   "],
    [" _ ",
     " _|",
     " _|",
     "   "],
    ["   ",
     "|_|",
     "  |",
     "   "],
    [" _ ",
     "|_ ",
     " _|",
     "   "],
    [" _ ",
     "|_ ",
     "|_|",
     "   "],
    [" _ ",
     "  |",
     "  |",
     "   "],
    [" _ ",
     "|_|",
     "|_|",
     "   "],
    [" _ ",
     "|_|",
     " _|",
     "   "]
  ]

  def self.convert_group(lines)
    raise ArgumentError, "Invalid number of lines" unless lines.size == LINES_PER_GROUP
    lines.map do |line|
      line.chars.each_slice(COLUMNS_PER_DIGIT).to_a
    end.transpose.map do |digit|
      convert_digit(digit.map(&:join))
    end.join
  end

  def self.convert_digit(digit_lines)
    raise ArgumentError, "Invalid number of columns" unless digit_lines.all? { _1.length == COLUMNS_PER_DIGIT }
    i = NUMBERS.index(digit_lines)
    i.nil? ? "?" : i.to_s
  end
end
